### ISSUE TITLE
Regression causing errors during reflection for 1.4 FutureEngine

### ISSUE THREAD

**Issue by matthewgdv** [sqlalchemy-redshift/sqlalchemy-redshift#269]
Just copying the description over from where I raised this on the [Spectrum PR](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/263):

Hi all, just wanted to mention a regression added in this PR that has forced me to pin `sqlalchemy-redshift==0.8.12` in my organization's library code repo. This happens specifically when using a `future=True` engine in sqlalchemy1.4, which previously worked fine. This is not the same as using sqlalchemy2.0, which has made large code changes to its connection and reflection logic relative to 1.4 (even with `FutureEngine`) that break `sqlalchemy-redshift` in other ways.

Basically, this PR added usages of `connection.connect()`  in the reflection code without checking if the connection was already open, which against a `FutureEngine` causes an error because connection forking is no longer allowed from sqlalchemy2.0 onwards (including a 1.4 `FutureEngine`).

Here is the offending line in my stack trace, though there may be others:

https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/blob/8e50cbee590c848fb8aad06013fca2dab365912f/sqlalchemy_redshift/dialect.py#L1128

**Comment by Brooke-white** [comment]
thanks for opening this matthew, appologies for the regression making it to production. starting to look into a fix for this.

**Comment by Brooke-white** [comment]
our current test suite is not sqlalchemy 2.0 ready. when making a fix for this issue, i'd like to be able to run our full integration test suite. Once it's migrated for future support, a fix will be made. Working on this migration now.

**Comment by matthewgdv** [comment]
Hiya, thanks for looking at it, and no worries, it happens. The whole `future=` opt-in toggle is an unusual thing for a library to offer and I guess we never had `FutureEngine`s in our test suite.

Iirc, back when 1.4 released there were a few PRs (at least one which I remember doing, but I vaguely recall seeing others on the issue tracker) to add compatibility with the 1.4 `future=` opt-in toggle. I think the compatibility for that is basically already there, it just needs this particular connection forking issue fixed.

I can totally understand that if there's already a push to migrate the project to full 2.0 it feels like wasted effort to fix this edge-case independently of the full migration, which would then use a 2.0-ready test-suite. 

I'd suggest that whether or not it's worth doing basically depends on the timescale of the 2.0 migration. If it looks like it might take something like 6+ months then it may be worth just hammering this out now. If it looks more like a timescale of 0-3 months then probably not.

Somewhat unrelated, but I've actually already made some attempts to run `sqlalchemy-redshift` SQLAlchemy 2.0 and one of the scariest errors I've uncovered is the use of the postgres `array_agg` function in their updated reflection logic. Since neither this function nor the array data type it returns exist in Redshift, we're probably going to have to completely rewrite whole chunks of the reflection logic that we were simply inheriting before.

**Comment by Brooke-white** [comment]
>  I think the compatibility for that is basically already there, it just needs this particular connection forking issue fixed.
this is consistent with what I'm seeing. It seems like it's really just that one line you pointed out wrt the connection.

Yep! Yesterday, I was able to get our test suite mostly running with future=True with a bunch of small compatibility changes (like using `sa.text`). I'd like to add a configuration in our CI to run with sqlalchemy 1.4 `future` toggled to try and guard against unintentional breaking changes to the FutureEngine support while we work on 2.0 support

I'm hoping to post PR with the fix for this + another regression by end of next week and will release once it merges.

Thanks for sharing your findings around `array_agg` -- that helps to know there's at least one aspect of the migration that will require some lifting on sqlalchemy-redshift's end. Redshift's relationship with arrays should make supporting 2.0 very interesting! 😄

**Comment by matthewgdv** [comment]
Just wanted to say, thanks a bunch for taking care of this so quickly. I really appreciate it :)
